rework http error handling closes #32

p-mcgowan opened the pull request:
Implement a generic, extensible HttpErrorException which can handle all HTTP error types with a single middleware handler.

Somewhat backwards compatible - in order to preserve the previous custom-error behaviour while still handling errors in the worker, the call sign of old http errors has changed to a constuctor pattern `new (...args: any[]) => Class`

With a few small changes, was able to get this up and running using another project's tests - after deleting most of the http folder and recreating it with the new errors, then replacing any reference to the custom-errors with the new patter, all tests in that particular project still passed.

Also added some index files to clean up imports and exports.

closes #32

Acrontum-Carmichael commented:
So as discussed - i implemented the return default and things build ok.. but.. now i just moved the problem from the domain logic to the tests.. 

issue - it will break any tests where the test check the e is an instance of
Does not break the code though..

The middleware will also complain when people start pulling this new tpl code

p-mcgowan commented:
yeah looks better - which tests will it break? existing tests that use `instanceof http404` for example will break, as `http404` is now a `NotFoundException`, but new tests or changes to domains should use the new pattern and also check for those typings.

I'm not sure how it will all affect the worker, but I think maybe we should throw together a test suite for both worker-handled and not worker handled, and see if we can put something together that works for both - what do you think?

p-mcgowan commented:
ah yeah, I forgot about that too - I think it would be good to split out generated / replaceable code from user-editable code.

For example, the strategy with the exceptions is the the base class on it's own can be completely overwritten and updated, so long as it still makes a call to the user-editable error formatting code. by splitting the two concerns, we can still make backwards-compatible changes while not breaking anything.

In the case of the middleware, we could to the exact same thing - have the default middleware be completely replaceable, but also allow for user middleware to be plugged in and modified. it's slightly more complicated due to the fact that order is important, but makes for easier rolling updates and a clear separation between auto-generated and auto-updated / overwritten files vs ones which are simply scaffolded once, but not overridden.

For the most part, that's the way it currently is, and most of the code in the http/ folder will be auto-updated, but there are still a few files that don't get overridden that cause these issues. It's also not always clear which files will and wont be updated.

So likely out of the scope of this merge, but it would be nice if we could implement a plugin-style system here with the http layer, so that we can make breaking changes in the future (or, more accurately, make changes without breaking user code).

p-mcgowan commented:
for example with the middleware - 
if we remove the 2 middleware files in `http/nodegen`, then create `http/nodegen/middleware/applicationMiddleware.ts` as such:
```typescript
// http/nodegen/middleware/applicationMiddleware.ts
import bodyParser from 'body-parser';
import express from 'express';
import expressFormData from 'express-form-data';
import morgan from 'morgan';
import { tmpdir } from 'os';
import requestIp from 'request-ip';
import packageJson from '../../package.json';
import { handleDomain404, handleExpress404, handleHttpException, handleValidationErrors } from './nodegen/middleware/';
import corsMiddleware from './nodegen/middleware/corsMiddleware';
import headersCaching from './nodegen/middleware/headersCaching';
import queryArrayParserMiddleware from './nodegen/middleware/queryArrayParserMiddleware';

export const responseHeaders = (app: express.Application): void => {
  app.use(corsMiddleware());
  app.use(headersCaching());
};

export const requestParser = (app: express.Application): void => {
  // parse data with connect-multiparty
  app.use(
    expressFormData.parse({
      autoClean: true,
      autoFiles: true,
      uploadDir: tmpdir(),
    })
  );
  app.use(bodyParser.json({ limit: '50mb' }));

  // parse query params
  app.use(queryArrayParserMiddleware());

  // clear all empty files (size == 0)
  app.use(expressFormData.format());

  // union body and files
  app.use(expressFormData.union());

  // parse the body
  app.use(bodyParser.urlencoded({ extended: false }));
  app.use(bodyParser.json());

  app.use(requestIp.mw());
};

export const accessLogger = (app: express.Application): void => {
  // Log all requests
  app.use(
    morgan(
      `[${packageJson.name}] :remote-addr [:date[clf]] ":method :url HTTP/:http-version" :status :res[content-length]`
    )
  );
};

/**
 * Injects routes into the passed express app
 * @param app
 */
export const requestMiddleware = (app: express.Application): void => {
  accessLogger(app);
  requestParser(app);
  responseHeaders(app);
};

/**
 * Injects routes into the passed express app
 * @param app
 */
export const responseMiddleware = (app: express.Application): void => {
  app.use(handleExpress404());
  app.use(handleDomain404());

  // Validation requests
  app.use(handleValidationErrors());

  // Handle HTTP errors
  app.use(handleHttpException());
};

```
Then in `app.ts` (from a fresh generation):
```typescript
import express from 'express'
import middlewaresImporter from './http/middlewaresImporter'
import routesImporter from './http/nodegen/routesImporter'
import { applicationMiddleware } from '@http/middleware'

/**
 * Returns a promise allowing the server or cli script to know
 * when the app is ready; eg database connections established
 */
export default (): Promise<express.Express> => {
  return new Promise((resolve, reject) => {
    // Here is a good place to connect to databases if required,
    // resolve once connected else reject
    const app = express()

    applicationMiddleware.requestMiddleware(app)
    routesImporter(app)
    applicationMiddleware.responseMiddleware(app)
    return resolve(app)
  })
}
```
this allows the nodegen folder to be recreated / updated without breaking user code in app.ts - it also allows for greater flixibility - the user can now modify default middleware (by including or excluding it), or write and apply their own (which is already supported to some degree, but right now there is a mix of replaceable and user edited code)

p-mcgowan commented on src/http/nodegen/middleware/handle422.ts:
```diff
@@ -9,6 +9,10 @@ import NodegenRequest from '../../interfaces/NodegenRequest';
 export default () => {
+console.warn(
+    `Deprecation warning: handle422() and the 422.ts error will be removed from the codebase in the future.
+Please use GoneException and allow it to be caught by the handleHttpException.ts
```
Unprocessable

Acrontum-Carmichael commented:
I have given this a good old test this evening and all seems to be working nicely so far @p-mcgowan I already merged the other fix to dev then master and merged dev into this branch. Cannot get it crash. I also pulled this into a random MS.. so full backward is good.. i would say this one is good to launch

p-mcgowan commented:
Awesome, Ok then I will start the merge and publish process at some point today
